File: Kdiff.py
# Time Complexity : O(n) where n is the number of elements in the list
# Space Complexity : O(m) where m is the number of unique elements in the list
# Did this code successfully run on LeetCode :  yes
# Any problem you faced while coding this : no



class Solution:
    def findPairs(self, nums: List[int], k: int) -> int:

    #when target is negative, we wil return 0
        if k < 0:
            return 0

        hashMap = {}
        output = 0

    #we iterate through the list and keep track of the frequency of numbers
        for num in nums:

            if num in hashMap:

                hashMap[num] += 1
            else:

                hashMap[num] = 1

        if k == 0:

        #If target is zero we will return number of items that have frequecy greater than 1
            for key, value in hashMap.items():

                if value > 1:

                    output += 1

            return output

        else:

        #else we will return the number of (element + k) present in the hashmap
            for key in hashMap:

                if key + k in hashMap:

                    output += 1

            return output

File: k_diff.cpp
// Time Complexity :O(n)
// Space Complexity :O(1)
// Did this code successfully run on Leetcode :yes
// Any problem you faced while coding this :

/*
In this approach we will use the hashmap for the optimal solution, we will first insert 
all the elements into the hashmap and insert the frequency as values.

now we will iterate through the hashmap and find the key+target value in the hashmap
which is basically the difference. In this approach we will basically always be finding the 
bigger element, from the which the current element will be subtracted will give the 
target.

Now, while iterating through the hashmap, if the target is 0, 2 elements need to
be of the same key(ex,(2,2), 3,3 etc), so for each key we will check if their frequency is
greater than 1, then count++.

If the target is a non zero value, we will be finding the complement of that value in the
hashmap and incrementing the count.

finally return the count.



*/


// Your code here along with comments explaining your approach

#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    int findPairs(vector<int>& nums, int k) {

        unordered_map<int,int> map;
        int count=0;

        for(int i=0;i<nums.size();i++){
            map[nums[i]]++;

        }

        for(auto p:map){
            if((k==0)&&p.second>1){
                count++;
            }

            if(k!=0 && map.find(k+p.first)!=map.end()){
                count++;
            }
        }

        return count;

        
        
    }
};